### PR TITLE
JA: add site/ja/tutorials/text/transformer.ipynb

### PR DESCRIPTION
**Pull request by masa-ita**

### CODE REVIEW

File: site/ja/tutorials/text/transformer.ipynb [hunk 1]
```diff
@@ -0,0 +1,2077 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "aOz-4_XIhaTP"
+   },
+   "source": [
+    "すべてのクエリをまとめる。"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: [weak suggestion]
```
すべてのクエリをまとめる。
=====>
すべてのクエリをまとめます。
```

File: site/ja/tutorials/text/transformer.ipynb [hunk 2]
```diff
@@ -0,0 +1,2077 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "0D8FJue5lDyZ"
+   },
+   "source": [
+    "試しに、`MultiHeadAttention`レイヤーを作ってみましょう。シーケンス`y`の各位置において、`MultiHeadAttention`はシーケンスのその位置以外の全ての位置に対して8つのアテンション・ヘッドを実行し、各位置それぞれで同じ長さの新しいベクトルを返します。"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: [suggestion]
```
シーケンス`y`の各位置において、`MultiHeadAttention`はシーケンスのその位置以外の全ての位置に対して
8つのアテンション・ヘッドを実行し、各位置それぞれで同じ長さの新しいベクトルを返します。
=====>
シーケンス `y` の各位置において、`MultiHeadAttention` はシーケンスのすべての位置に対して8つのヘッドを用いて
アテンションを計算し、各位置それぞれで同じ長さの新しいベクトルを返します。
```

For each location `i`, if `MultiHeadAttention` runs attention heads all other locations excepting the location `i`, `MultiHeadAttention` returns a new vector of shorter length.

File: site/ja/tutorials/text/transformer.ipynb [hunk 3]
```diff
@@ -0,0 +1,2077 @@
...
+    "colab_type": "text",
+    "id": "JPmbr6F1C-v_"
+   },
+   "source": [
+    "各マルチヘッド・アテンション・ブロックは3つの入力：Q(query), K(key), V(value)を取ります。\n",
+    "これらは、線形（Dense）レイヤーを通され、マルチヘッドに分割されます。\n",
+    "\n",
+    "上記で定義した`scaled_dot_product_attention`は（効率のためにブロードキャストで）各ヘッドに適用されます。アテンション・ステップにおいては、適切なマスクを使用しなければなりません。その後、各ヘッドのアテンション出力は（`tf.transpose`と`tf.reshape`を使って）結合され、最後の`Dense`レイヤーに通されます。\n",
+    "\n",
+    "単一のアテンション・ヘッドのかわりに、Q、K、およびVは複数のヘッドに分割されます。なぜなら、それによって、モデルが異なる表現空間の異なる位置の情報について、連携して注意することができるからです。分割後は、各ヘッドの次元が小さくなるので、全体の計算コストは、全ての次元を持つ単一のアテンション・ヘッドと同じです。"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: [weak suggestion]
```
連携して注意することができるからです。
=====>
連携してアテンションを計算できるからです。
```

File: site/ja/tutorials/text/transformer.ipynb [hunk 4]
```diff
@@ -0,0 +1,2077 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "MfYJG-Kvgwy2"
+   },
+   "source": [
+    "Transformerモデルは、標準の[アテンション付きシーケンス・トゥー・シーケンスモデル](nmt_with_attention.ipynb)と同じ一般的パターンを踏襲します。\n",
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: [weak suggestion]
```
一般的パターン
=====>
一般的なパターン
```

File: site/ja/tutorials/text/transformer.ipynb [hunk 5]
```diff
@@ -0,0 +1,2077 @@
...
+   },
+   "source": [
+    "### エンコーダー・レイヤー\n",
+    "\n",
+    "それぞれのエンコーダー・レイヤーは次のようなサブレイヤーから成っています。\n",
+    "\n",
+    "1.  マルチヘッド・アテンション（パディング・マスク付き）\n",
+    "2.  ポイントワイズ・フィードフォワード・ネットワーク\n",
+    "\n",
+    "サブレイヤーにはそれぞれ残差接続があり、その後にレイヤー正規化が続きます。残差接続は、深いネットワークでの勾配消失問題を回避するのに役立ちます。\n",
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: [weak suggestion]

```
残差接続
=====>
残差コネクション
```

It seems to be the standard way to translate the word `residual` into `残差`. However, I guess it is a little unnatural to translate `residual connection` into `残差接続`. I think it is better to leave the word `connection` as is.

File: site/ja/tutorials/text/transformer.ipynb [hunk 6]
```diff
@@ -0,0 +1,2077 @@
...
+    "colab_type": "text",
+    "id": "6LO_48Owmx_o"
+   },
+   "source": [
+    "### デコーダー・レイヤー\n",
+    "\n",
+    "各デコーダー・レイヤーは次のようなサブレイヤーからなります。\n",
+    "\n",
+    "1. マスク付きマルチヘッド・アテンション（ ルックアヘッド・マスクおよびパディング・マスク付き）\n",
+    "2. （パディング・マスク付き）マルチヘッド・アテンション。V(value)とK(key)は*エンコーダー出力*を入力そして受け取る。Q(query)は*マスク付きマルチヘッド・アテンション・サブレイヤー*の出力を受け取る。\n",
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: [suggestion]
```
V(value)とK(key)は*エンコーダー出力*を入力そして受け取る。
=====>
V (value) と K (key) は*エンコーダーの出力*を入力として受け取る。
```

File: site/ja/tutorials/text/transformer.ipynb [hunk 7]
```diff
@@ -0,0 +1,2077 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "zVjWCxFNcgbt"
+   },
+   "source": [
+    "このサンプルを小さく、比較的高速にするため、 *num_layers, d_model, and dff*の値は小さくされています。\n",
+    "\n",
+    "Transformerのベースモデルで使われている値は*num_layers=6*, *d_model = 512*, *dff = 2048*です。 Transformerの他のバージョンについては、[論文](https://arxiv.org/abs/1706.03762)を参照してください。\n",
+    "\n",
+    "Note: 下記の値を変更することで、さまざまなタスクでSoTA（訳注：State of The Art、その時点での最高性能）を達成したモデルが得られます。"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: Nice translation 🎉

File: site/ja/tutorials/text/transformer.ipynb [hunk 8]
```diff
@@ -0,0 +1,2077 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "Fzuf06YZp66w"
+   },
+   "source": [
+    "チェックポイントのパスとチェックポイント・マネージャーを作成。これは、`n`エポックごとにチェックポイントを保存するのに使用される。"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: [suggestion]
```
チェックポイントのパスとチェックポイント・マネージャーを作成。
これは、`n`エポックごとにチェックポイントを保存するのに使用される。
=====>
チェックポイントのパスとチェックポイント・マネージャーを作成します。
これは、`n`エポックごとにチェックポイントを保存するのに使用されます。
```

File: site/ja/tutorials/text/transformer.ipynb [hunk 9]
```diff
@@ -0,0 +1,2077 @@
...
+    "\n",
+    "* ポルトガル語のトークナイザー(`tokenizer_pt`)を使用して入力文をエンコードします。さらに、モデルの訓練に使用されたものと同様に、開始および終了トークンを追加します。これが、入力のエンコードです。\n",
+    "* デコーダーの入力は、`start token == tokenizer_en.vocab_size`です。\n",
+    "* パディング・マスクとルックアヘッド・マスクを計算します。\n",
+    "* `decoder`は、`encoder output`と自分自身の出力（セルフアテンション）を見て、予測値を出力します。\n",
+    "* 最後の単語を選択し、そのargmaxを計算します。\n",
+    "* デコーダーの入力に予測された単語を結合し、デコーダーに渡します。\n",
+    "* このアプローチでは、デコーダーは自分自身が予測した過去の単語にもとづいて次の単語を予測します。\n",
+    "\n",
+    "Note: ここで使われているモデルは、サンプルをより速くするために容量が少ないものであるため、予測はあまり正確ではありません。論文の結果を再現するには、データセット全体を使用し、上記のハイパーパラメーターを変更して、ベースのTransformerモデルまたはTransformer XLを使用します。"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/transformer.ipynb]: [suggestion]
```
サンプルをより速くするために容量が少ないものであるため、
=====>
より早く実行できるようにした能力の低いものであるため、
```

File: site/ja/tutorials/text/transformer.ipynb [hunk 10]
```diff
@@ -0,0 +1,2077 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "Tx1sFbR-9fRs"
+   },
+   "source": [
+    "`.map()` の中の演算はグラフモードで実行され、numpy 属性を持たないグラフテンソルを受け取ります。`tokenizer` は、整数にエンコードを行うため、文字列あるいは Unicode シンボルを受け取ることを期待しています。したがって、このエンコーディングは、文字列を値として含む numpy 属性を持った Eager テンソルを受け取る `tf.py_function` の中で実行する必要があります。"
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks．Deleted.

File: site/ja/tutorials/text/transformer.ipynb [hunk 11]
```diff
@@ -0,0 +1,2077 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "nBQuibYA4n0n"
+   },
+   "source": [
+    "## 位置エンコーディング\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thank you.  But the word is used in a lot of articles.

File: site/ja/tutorials/text/transformer.ipynb [hunk 12]
```diff
@@ -0,0 +1,2077 @@
...
+    "\n",
+    "このモデルには再帰や畳込みが含まれないので、モデルに文中の単語の相対的な位置の情報を与えるため、位置エンコーディングを追加します。\n",
+    "\n",
+    "位置エンコーディングベクトルは埋め込みベクトルに加算します。埋め込みはトークンをd次元空間で表現します。そこでは、同じような意味を持つトークンが近くに位置することになります。しかし、埋め込みは単語の文中の相対的位置をエンコードしません。したがって、位置エンコーディングを加えることで、単語は、d次元空間の中で、*意味と文中の位置の近さ*にもとづいて近くに位置づけられます。\n",
+    "\n",
+    "もう少し知りたければ [位置エンコーディング](https://github.com/tensorflow/examples/blob/master/community/en/position_encoding.ipynb) のノートブックを参照してください。位置エンコーディングを計算する式は下記のとおりです。\n",
+    "\n",
+    "$$\\Large{PE_{(pos, 2i)} = sin(pos / 10000^{2i / d_{model}})} $$\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: That's strange.  Is it a problem of GitHub?

File: site/ja/tutorials/text/transformer.ipynb [hunk 13]
```diff
@@ -0,0 +1,2077 @@
...
+    "id": "vsxEE_-Wa1gF"
+   },
+   "source": [
+    "<img src=\"https://www.tensorflow.org/images/tutorials/transformer/scaled_attention.png\" width=\"500\" alt=\"scaled_dot_product_attention\">\n",
+    "\n",
+    "Transformerで使われているアテンション関数は3つの入力；Q(query), K(key), V(value)を取ります。このアテンションの重みの計算に使われている式は下記の通りです。\n",
+    "\n",
+    "$$\\Large{Attention(Q, K, V) = softmax_k(\\frac{QK^T}{\\sqrt{d_k}}) V} $$\n",
+    "\n",
+    "内積アテンションは、深度の平方根をファクターとしてスケールされています。これは、深度が大きくなると、内積が非常に大きくなり、勾配が小さい部分のsoftmax関数が非常に厳しくなるためです。\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  I agree.

File: site/ja/tutorials/text/transformer.ipynb [hunk 14]
```diff
@@ -0,0 +1,2077 @@
...
+    "\n",
+    "例えば、`Q`と`K`が平均0分散1だと思ってください。これらの行列積は、平均0分散は`dk`となります。したがって、`Q`と`K`のマトリクス積が平均0分散1で有るべきであるため、（他の数字ではなく）*`dk`の平方根*がスケーリングに使われ、緩やかなソフトマックスが得られるのです。\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Yes.  It's better.

File: site/ja/tutorials/text/transformer.ipynb [hunk 15]
```diff
@@ -0,0 +1,2077 @@
...
+    "\n",
+    "マスクには、（負の無限大に近い）-1e9が掛けられています。これは、マスクがQとKのスケール済みマトリクス積と合計され、ソフトマックスの直前に適用されるからです。目的は、これらのセルをゼロにしてしまうことで、大きなマイナスの入力は、ゼロに近い出力となります。"
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  Changed.

File: site/ja/tutorials/text/transformer.ipynb [hunk 16]
```diff
@@ -0,0 +1,2077 @@
...
+    "  # matmul_qkをスケール\n",
+    "  dk = tf.cast(tf.shape(k)[-1], tf.float32)\n",
+    "  scaled_attention_logits = matmul_qk / tf.math.sqrt(dk)\n",
+    "\n",
+    "  # マスクをスケール済みテンソルに加算\n",
+    "  if mask is not None:\n",
+    "    scaled_attention_logits += (mask * -1e9)  \n",
+    "\n",
+    "  # softmax は最後の軸(seq_len_k)に沿って正規化し\n",
+    "  # 合計が1となるように\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  It's easier to understand.

File: site/ja/tutorials/text/transformer.ipynb (hunk 1 above)
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  Fixed.

File: site/ja/tutorials/text/transformer.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  It's better.

File: site/ja/tutorials/text/transformer.ipynb (hunk 2 above)
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Oh! So that's it.

File: site/ja/tutorials/text/transformer.ipynb (hunk 3 above)
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  It's more natural.

File: site/ja/tutorials/text/transformer.ipynb (hunk 4 above)
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  Fixed.

File: site/ja/tutorials/text/transformer.ipynb (hunk 5 above)
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Well.  Google search on "残差接続" returns more results compared to "残差コネクション".
So I'd like to leave the word as is.

File: site/ja/tutorials/text/transformer.ipynb (hunk 6 above)
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  Fixed.

File: site/ja/tutorials/text/transformer.ipynb (hunk 7 above)
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thank you.

File: site/ja/tutorials/text/transformer.ipynb (hunk 9 above)
Comment (masa-ita) [reply site/ja/tutorials/text/transformer.ipynb]: Thanks.  It's better.

File: site/ja/tutorials/text/transformer.ipynb (hunk 12 above)
[the remark this answers is not in the document]
Comment (AseiSugiyama) [reply site/ja/tutorials/text/transformer.ipynb]: Yes, I think GitHub has some issues about rendering Jupyter notebook.

File: site/ja/tutorials/text/transformer.ipynb (hunk 5 above)
Comment (AseiSugiyama) [reply site/ja/tutorials/text/transformer.ipynb]: Right, I checked it on Google search and was able to reproduce same result. I agree with you.

**Comment by AseiSugiyama** [comment]
LGTM!

**Comment by AseiSugiyama** [comment]
@lamberta This PR has been finished review process by the Japanese community. Could you check this?

**Pull request merged**